extensions/amp-pan-zoom/1.0/hooks/usePanZoomState.ts

```diff
@@ -0,0 +1,214 @@
+  if (contentSize.width * newScale > containerSize.width) {
+    [minX, maxX] = [maxX, minX]; // (swap)
+  }
+  if (contentSize.height * newScale > containerSize.height) {
+    [minY, maxY] = [maxY, minY]; // (swap)
+  }
+
+  const extent = allowExtent && newScale > 1 ? 0.25 : 0;
+
+  return {
+    posX: boundValueSpring(posX, minX, maxX, extent),
+    posY: boundValueSpring(posY, minY, maxY, extent),
+    scale: newScale,
+  };
+};
+
+function initReducer(config: PanZoomConfig): State {
+  const {
+    initialScale = DEFAULT_INITIAL_SCALE,
+    initialX = DEFAULT_ORIGIN,
+    initialY = DEFAULT_ORIGIN,
+    maxScale = DEFAULT_MAX_SCALE,
+  } = config;
+  return {
+    ...initialState,
+    posX: Number(initialX),
+    posY: Number(initialY),
+    scale: Number(initialScale),
+    maxScale: Number(maxScale),
+  };
+}
+
+type PickState<keys extends keyof State> = Required<Pick<State, keys>>;
+
+export function usePanZoomState(config: PanZoomConfig) {
+  const [state, setState] = useState(() => initReducer(config));
+  const actions = useMemo(
+    () => ({
+      updateBounds(
+        payload: PickState<'contentOffset' | 'containerSize' | 'contentSize'>
+      ) {
+        setState((state) => {
+          const newState = {
+            ...state,
+            ...payload,
+          };
+          return {
+            ...newState,
+            // Ensure the element is still in-bounds:
+            ...updateView(newState, newState),
+          };
+        });
+      },
+      draggingStart() {
+        setState((state) => {
+          return {
+            ...state,
+            isDragging: true,
+            allowExtent: true,
+          };
+        });
+      },
+      draggingRelease() {
+        setState((state) => {
+          const newState: State = {
+            ...state,
+            isDragging: false,
+            allowExtent: false,
+          };
+          // "Snap back" when we release:
+          const newPosition = updateView(newState, newState);
+          return {
+            ...newState,
+            ...newPosition,
+          };
+        });
+      },
+      updateScale(payload: {
+        anchorX?: number;
+        anchorY?: number;
+        scale?: number;
+      }) {
+        setState((state) => {
+          const {
+            anchorX = state.containerSize.width / 2,
+            anchorY = state.containerSize.height / 2,
+            scale = state.scale === state.maxScale
+              ? state.minScale
+              : Math.floor(state.scale) + 1,
+          } = payload;
+          return {
+            ...state,
+            ...updateView(
+              state,
+              updateScaleFromAnchor(state, anchorX, anchorY, scale)
+            ),
+          };
+        });
+      },
+      transform(payload: Partial<PickState<'posX' | 'posY' | 'scale'>>) {
+        setState((state) => ({
+          ...state,
+          ...updateView(state, payload),
+        }));
+      },
+    }),
+    [setState]
```

Comment:
Isn't `setState` a constant function identity? I think this `useMemo` can never be busted? Could this just be `[]` instead?

Reply:
Yep, I guess just old habits.  FIxed.